compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -5484,13 +5479,19 @@ public void visit(BLangQueryAction queryAction) {
         resultType = types.checkType(doClause.pos, actualType, expType, DiagnosticErrorCode.INCOMPATIBLE_TYPES);
         queryFinalClauses.pop();
         queryEnvs.pop();
-        prevEnvs.pop();
+        if (!breakToParallelQueryEnv) {
+            prevEnvs.pop();
+        }
     }
 
     @Override
     public void visit(BLangFromClause fromClause) {
         boolean prevBreakToParallelEnv = this.breakToParallelQueryEnv;
-        this.breakToParallelQueryEnv = true;
+        if (fromClause.collection.getKind() == NodeKind.QUERY_EXPR ||
+                (fromClause.collection.getKind() == NodeKind.GROUP_EXPR
+                        && ((BLangGroupExpr) fromClause.collection).expression.getKind() == NodeKind.QUERY_EXPR)) {
+            this.breakToParallelQueryEnv = true;
```

Comment:
Some places we have used `this.breakToParallelQueryEnv` but in other occasions we don't use `this`. Shall we make the code consistent in all the usages of this variable.